File: c++/gcd.cpp
#include<iostream>
using namespace std;

int egcd(int a, int b, int& resa, int& resb){
    // return resa * a + resb * b = gcd
    if(a == 0){
        resa = 0, resb = 1; return b;
    }
    int x1 = 0, y1 = 0;
    int gcd = egcd(b%a, a, x1, y1);
    resa = y1 - (b/a) * x1;
    resb = x1;
    return gcd;
}
int CRT(int n1, int n2, int a1, int a2){
    int m1, m2;
    egcd(n1, n2, m1, m2);
    int x = a1 * m2 * n2 + a2 * m1 * n1;
    return x;
}
int main(){
    int a, b;
    cout<<egcd(35, 6, a, b)<<" ";
    cout<<a<<" "<<b<<endl;
    cout<<CRT(35, 6, 17, 5)<<" ";
    return 0;
}

File: python/gcd.py
def egcd(a, b):
    # ref: https://www.geeksforgeeks.org/python-program-for-basic-and-extended-euclidean-algorithms-2/
    if a == 0 :
        return b,0,1
    gcd,x1,y1 = egcd(b%a, a)
    x = y1 - (b//a) * x1
    y = x1
    return gcd,x,y

def CRT(n1, n2, a1, a2):
    # obtain x (mod lcm(n1, n2))
    #        x = a1 (mod n1)
    #        x = a2 (mod n2)
    #
    # ref: https://en.wikipedia.org/wiki/Chinese_remainder_theorem#Case_of_two_moduli
    
    gcd, m1, m2 = egcd(n1, n2)
    x = a1 * m2 * n2 + a2 * m1 * n1
    x %= n1 * n2 // gcd
    return x

if __name__ == '__main__':
    print(egcd(35, 15))
    print(CRT(35, 6, 12, 7))